explorer/interpreter/type_checker.cpp

```diff
@@ -2335,10 +2336,28 @@ auto TypeChecker::MatchImpl(const InterfaceType& iface,
       *trace_stream_ << "matched with " << *impl.type << " as "
                      << *impl.interface << "\n\n";
     }
-    CARBON_ASSIGN_OR_RETURN(
-        const auto* subst_witness,
-        SubstituteCast<Witness>(*bindings_or_error, impl.witness));
-    return {subst_witness};
+
+    auto subst_witness_or_error =
+        SubstituteCast<Witness>(*bindings_or_error, impl.witness);
+
+    if (!subst_witness_or_error.ok()) {
+      std::string error;
+      llvm::raw_string_ostream stream(error);
```

Comment:
Marking unresolved. The implementation still creates a raw_string_ostream instead of using the ErrorBuilder stream directly.

Reply:
Thanks. I've reintroduced the issue after solving it because I didn't find a quick way to stream directly the variable number of bindings. I'll look into it

Reply:
auto note_stream = ErrorBuilder() << ...
...
return note_stream;